File: servers/go/internal/persistence/postgres/loan/write_repo.go
package loan

import (
	"context"
	"fmt"
	"time"

	domainbc "github.com/christophercaldwell/model-architecture/go/internal/domain/bookcopy"
	domain "github.com/christophercaldwell/model-architecture/go/internal/domain/loan"
	domainm "github.com/christophercaldwell/model-architecture/go/internal/domain/member"
	"github.com/jackc/pgx/v5"
)

type WriteRepo struct {
	tx pgx.Tx
}

func NewWriteRepo(tx pgx.Tx) *WriteRepo {
	return &WriteRepo{tx: tx}
}

func (r *WriteRepo) Create(ctx context.Context, prepared domain.LoanPrepared) (*domain.Loan, error) {
	const q = `
		WITH next_id AS (
			SELECT nextval(pg_get_serial_sequence('library.loan', 'loan_id'))::integer AS loan_id
		), inserted AS (
			INSERT INTO library.loan (loan_id, loan_ident, book_copy_id, member_id)
			OVERRIDING SYSTEM VALUE
			SELECT next_id.loan_id, 'LN-' || lpad(next_id.loan_id::text, 6, '0'), $1, $2
			FROM next_id
			RETURNING loan_id, loan_ident
		)
		SELECT inserted.loan_id, inserted.loan_ident FROM inserted`

	var loanID int32
	var loanIdent string
	err := r.tx.QueryRow(ctx, q, int32(prepared.BookCopyID), int32(prepared.MemberID)).Scan(&loanID, &loanIdent)
	if err != nil {
		return nil, fmt.Errorf("create loan: %w", err)
	}
	now := time.Now()
	l := domain.Loan{
		ID:         domain.LoanID(loanID),
		Ident:      domain.LoanIdent(loanIdent),
		DtCreated:  now,
		DtModified: now,
		BookCopyID: prepared.BookCopyID,
		MemberID:   prepared.MemberID,
		DtDue:      nil,
		DtReturned: nil,
	}
	return &l, nil
}

func (r *WriteRepo) End(ctx context.Context, id domain.LoanID) error {
	const q = `
		UPDATE library.loan l
		SET dt_returned = CURRENT_TIMESTAMP
		WHERE l.loan_id = $1`

	_, err := r.tx.Exec(ctx, q, int32(id))
	if err != nil {
		return fmt.Errorf("end loan: %w", err)
	}
	return nil
}

func (r *WriteRepo) FindActiveByBookCopyIDForUpdate(ctx context.Context, id domainbc.BookCopyID) (*domain.Loan, error) {
	const activeLoanCols = `l.loan_id, l.loan_ident, l.dt_created, l.dt_modified, l.book_copy_id, l.member_id,
		NULLIF(l.dt_due, '9999-01-01 00:00:00+00'::TIMESTAMPTZ),
		NULLIF(l.dt_returned, '9999-01-01 00:00:00+00'::TIMESTAMPTZ)`
	q := `SELECT ` + activeLoanCols + `
		FROM library.loan l
		WHERE l.book_copy_id = $1
		  AND l.dt_returned = '9999-01-01 00:00:00+00'::TIMESTAMPTZ
		ORDER BY l.loan_id DESC
		LIMIT 1
		FOR UPDATE OF l`

	row, err := scanLoanRow(r.tx.QueryRow(ctx, q, int32(id)))
	if err != nil {
		return nil, fmt.Errorf("find active loan by book copy id for update: %w", err)
	}
	if row == nil {
		return nil, nil
	}
	l := rowToDomain(*row)
	return &l, nil
}

func (r *WriteRepo) CountActiveByMemberID(ctx context.Context, id domainm.MemberID) (int64, error) {
	const q = `
		SELECT COUNT(*)::BIGINT
		FROM library.loan l
		WHERE l.member_id = $1
		  AND l.dt_returned = '9999-01-01 00:00:00+00'::TIMESTAMPTZ`

	var count int64
	if err := r.tx.QueryRow(ctx, q, int32(id)).Scan(&count); err != nil {
		return 0, fmt.Errorf("count active loans by member id (write): %w", err)
	}
	return count, nil
}

File: server/crates/persistence/src/loan/write_repo.rs
use anyhow::{Context, Result};
use chrono::Utc;
use domain::{
    book_copy::BookCopyId,
    loan::{Loan, LoanId, LoanIdent, LoanPrepared},
    member::MemberId,
};
use sqlx::PgExecutor;

use crate::loan::read_repo::{CountDbRow, LoanDbRow};

#[derive(sqlx::FromRow)]
pub struct LoanPreparedResult {
    pub loan_id: i32,
    pub loan_ident: String,
}

pub(crate) async fn create_loan_with(
    executor: impl PgExecutor<'_>,
    insert: &LoanPrepared,
) -> Result<Loan> {
    let prepared_result = sqlx::query_file_as!(
        LoanPreparedResult,
        "sql/loan/commands/create.sql",
        insert.book_copy_id.0,
        insert.member_id.0,
    )
    .fetch_one(executor)
    .await
    .context("Failed to create loan")?;

    let now = Utc::now();
    let created_loan = Loan {
        id: LoanId(prepared_result.loan_id),
        ident: LoanIdent(prepared_result.loan_ident),
        dt_created: now,
        dt_modified: now,
        book_copy_id: insert.book_copy_id,
        member_id: insert.member_id,
        dt_due: None,
        dt_returned: None,
    };
    Ok(created_loan)
}

pub(crate) async fn end_loan_with(executor: impl PgExecutor<'_>, id: LoanId) -> Result<()> {
    sqlx::query_file!("sql/loan/commands/end.sql", id.0,)
        .execute(executor)
        .await
        .context("Failed to end loan")?;

    Ok(())
}

pub(crate) async fn find_active_loan_by_book_copy_id_for_update_with(
    executor: impl PgExecutor<'_>,
    id: BookCopyId,
) -> Result<Option<Loan>> {
    let row = sqlx::query_file_as!(
        LoanDbRow,
        "sql/loan/commands/find_active_by_book_copy_id_for_update.sql",
        id.0
    )
    .fetch_optional(executor)
    .await
    .context("Failed to find active loan by book copy id")?;

    row.map(Loan::try_from).transpose()
}

pub(crate) async fn count_active_loans_by_member_id_with(
    executor: impl PgExecutor<'_>,
    id: MemberId,
) -> Result<i64> {
    let row = sqlx::query_file_as!(
        CountDbRow,
        "sql/loan/commands/count_active_by_member_id.sql",
        id.0
    )
    .fetch_one(executor)
    .await
    .context("Failed to count active loans by member id")?;

    Ok(row.count)
}
